lib/ansible/modules/net_tools/ldap/ldap_attrs.py

```diff
@@ -0,0 +1,340 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2019, Maciej Delmanowski <drybjed@gmail.com>
+# Copyright: (c) 2017, Alexander Korinek <noles@a3k.net>
+# Copyright: (c) 2016, Peter Sagerson <psagers@ignorare.net>
+# Copyright: (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+
+DOCUMENTATION = r'''
+---
+module: ldap_attrs
+short_description: Add or remove multiple LDAP attribute values.
+description:
+  - Add or remove multiple LDAP attribute values.
+notes:
+  - This only deals with attributes on existing entries. To add or remove
+    whole entries, see M(ldap_entry).
+  - The default authentication settings will attempt to use a SASL EXTERNAL
+    bind over a UNIX domain socket. This works well with the default Ubuntu
+    install for example, which includes a cn=peercred,cn=external,cn=auth ACL
+    rule allowing root to modify the server configuration. If you need to use
+    a simple bind to access your server, pass the credentials in I(bind_dn)
+    and I(bind_pw).
+  - For I(state=present) and I(state=absent), all value comparisons are
+    performed on the server for maximum accuracy. For I(state=exact), values
+    have to be compared in Python, which obviously ignores LDAP matching
+    rules. This should work out in most cases, but it is theoretically
+    possible to see spurious changes when target and actual values are
+    semantically identical but lexically distinct.
+version_added: '2.10'
+author:
+  - Jiri Tyr (@jtyr)
+  - Alexander Korinek (@noles)
+requirements:
+  - python-ldap
+options:
+  state:
+    required: false
+    type: str
+    choices: [present, absent, exact]
+    default: present
+    description:
+      - The state of the attribute values. If C(present), all given
+        values will be added if they're missing. If C(absent), all given
+        values will be removed if present. If C(exact), the set of values
+        will be forced to exactly those provided and no others. If
+        I(state=exact) and I(value) is empty, all values for this
+        attribute will be removed.
+  attributes:
+    required: true
+    type: dict
+    description:
+      - The attribute(s) and value(s) to add or remove. The complex argument format is required in order to pass
+        a list of strings (see examples).
+  ordered:
+    required: false
+    type: bool
+    default: 'no'
+    description:
+      - If C(yes), prepend list values with X-ORDERED index numbers in all
+        attributes specified in the current task. This is useful mostly with
+        I(olcAccess) attribute to easily manage LDAP Access Control Lists.
+  params:
+    description:
+    - Additional module parameters.
```

Comment:
Can you explain what this option does and why it is needed?

From the implementation, it looks like this can be used to overwrite arbitrary module options, without any type checks or other sanity checks. I don't think modules should have such options.

Reply:
This parameter can be used to [define connection options to the LDAP directory separate from the task itself](https://docs.debops.org/en/master/ansible/roles/debops.ldap/defaults/main.html#envvar-ldap__admin_auth_params). Since different connection methods might need different parameters, it's hard to implement this in a read-only playbook such as DebOps without using a workaround like this. It also gives the users a specific inventory variable to modify if their connection method is different than the default one. The parameter itself comes from the original implementation and is used in the `ldap_entry` module as well.

Perhaps you could reconsider implementing something like this officially for specific Ansible modules, so that the type and sanity checks could be provided? I reckon that many modules that interact with external resources (databases, devices, etc.) with different authentication methods could benefit from this.

Reply:
Ansible provides two alternatives for this:
 1. (the more tedious one) provide for every such module option a variable and specify it to every LDAP task in the role via `option_name: "{{ option_name | default(omit) }}"`. Then users can define the options they want, and only these will be passed on to all LDAP tasks. You can potentially save some complexity by using [some YAML tricks](https://docs.ansible.com/ansible/latest/user_guide/playbooks_advanced_syntax.html#yaml-anchors-and-aliases-sharing-variable-values).
 2. use [module defaults](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html). Adding a [module defaults group](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups) for LDAP modules would simplify this (if there are multiple LDAP modules which are used together).

Reply:
YAML anchors are probably more "native" way to do this, although I'm not sure how they work across the Ansible inventory, role default variables and role dependent variables. I would need to research that.

The module defaults might not be a good solution in niche cases, for example managing the OpenLDAP server `cn=config` database requires connecting via the `root` account, and when managing the normal LDAP directory, you would want to bind using a normal admin account over TLS. In such case the module defaults would probably clash. And from the looks of it, this is a playbook-level feature - for my use case it would have to work at least via the Ansible inventory since the playbooks are also meant to be read-only. Defining module defaults via role default variables looks to be not feasible either.

Still, since the `params` option is already present in the `ldap_entry` module, I hope it's not a blocker for merging the `ldap_attrs` module? The functionality could be removed from both modules later if a consensus is reached that it a problematic feature to have in Ansible core. Although it's currently perfect in DebOps case where playbooks and roles are read-only, so it would be sad to see it go.

Reply:
> YAML anchors are probably more "native" way to do this, although I'm not sure how they work across the Ansible inventory, role default variables and role dependent variables. I would need to research that.

They probably work only in one YAML file. So one can use them locally (in one role's task file), but not accross different YAML files.

> The module defaults might not be a good solution in niche cases, for example managing the OpenLDAP server `cn=config` database requires connecting via the `root` account, and when managing the normal LDAP directory, you would want to bind using a normal admin account over TLS. In such case the module defaults would probably clash. And from the looks of it, this is a playbook-level feature - for my use case it would have to work at least via the Ansible inventory since the playbooks are also meant to be read-only. Defining module defaults via role default variables looks to be not feasible either.

That's true. You can still combine the options from variables (which can come from inventory), though, AFAIK. And you can overwrite the module defaults per task with something else.

> Still, since the `params` option is already present in the `ldap_entry` module, I hope it's not a blocker for merging the `ldap_attrs` module?

I won't merge this PR because of it, but that doesn't mean others are as picky as me :) It would probably be best to discuss this in the Ansible project meeting, to see if such an option would be acceptable for the core team. (If I know it is, I would also merge this :) )

> The functionality could be removed from both modules later if a consensus is reached that it a problematic feature to have in Ansible core. Although it's currently perfect in DebOps case where playbooks and roles are read-only, so it would be sad to see it go.

Removing features is always a lot harder than adding them, so it's better to discuss this in more detail (with more people) before merging.

The best way (if you don't want to use module defaults) is probably to work with variables and `{{ xxx | default(omit) }}`. That also gives more fine control which module options can be overwritten by inventory, and which not. (And by which variables they can be overwritten in which situation.)